clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java

```diff
@@ -0,0 +1,195 @@
+        String contentSha256 = BinaryUtils.toHex(contentDigest);
+
+        HeaderSigningResult headerSigningResult = computeSignature(
+                request,
+                dateStamp,
+                timeStamp,
+                ALGORITHM,
+                contentSha256,
+                sanitizedCredentials);
+
+        request.addParameter("X-Amz-Signature", BinaryUtils.toHex(headerSigningResult.getSignature()));
+
+        return request;
+    }
+}
+
+public class AWSLakeFSTokenProvider implements LakeFSTokenProvider {
+    STSGetCallerIdentitPresigner stsPresigner;
+    AWSCredentialsProvider awsProvider;
+    String lakeFSAuthToken = null;
+    String stsEndpoint;
+    Map<String,String> stsAdditionalHeaders;
+    Long stsExpirationInSeconds;
+    AWSLakeFSTokenProvider(AWSCredentialsProvider awsProvider, STSGetCallerIdentitPresigner stsPresigner, String stsEndpoint, Map<String,String> stsAdditionalHeaders, Long stsExpirationInSeconds) {
+        this.stsPresigner = stsPresigner;
+        this.awsProvider = awsProvider;
+        this.stsEndpoint = stsEndpoint;
+        this.stsAdditionalHeaders = stsAdditionalHeaders;
+        this.stsExpirationInSeconds = stsExpirationInSeconds;
+    }
+
+    @Override
+    public String getToken() {
+        if (needsNewToken()) {
+            refresh();
+        }
+        return this.lakeFSAuthToken;
+    }
+
+    private boolean needsNewToken() {
+        return this.lakeFSAuthToken == null;
+    }
+
+    public String newPresignedGetCallerIdentityToken() throws Exception {
+        GeneratePresignGetCallerIdentityRequest stsReq = new GeneratePresignGetCallerIdentityRequest(
+                new URI(this.stsEndpoint),
+                this.awsProvider.getCredentials(),
+                this.stsAdditionalHeaders,
+                this.stsExpirationInSeconds
+        );
+        Request<GeneratePresignGetCallerIdentityRequest> signedRequest = this.stsPresigner.presignRequest(stsReq);
+
+        // generate anonymous class that will be seriallized to json string
+        JSONObject identityTokenParams = new JSONObject();
+        identityTokenParams.put("method", signedRequest.getHttpMethod().name());
+        identityTokenParams.put("endpoint", signedRequest.getEndpoint().toString());
+        identityTokenParams.put("signedHeaders", signedRequest.getHeaders().keySet().toArray());
+        identityTokenParams.put("expiration", this.stsExpirationInSeconds);
+        identityTokenParams.put("signedParams", signedRequest.getParameters());
+
+        // base64 encode
+        return "lakefs.iam.v1."+Base64.encodeBase64String(identityTokenParams.toString().getBytes());
+    }
+
+    private void newToken() throws Exception {
+        String token = this.newPresignedGetCallerIdentityToken();
+        // call lakeFS to exchange the token for a lakeFS token
+        this.lakeFSAuthToken = "lakefs jwt 123";
+    }
+
+    @Override
+    public void refresh() {
```

Comment:
I don't understand, it looks like this will _never_ refresh!  The first time it calls newToken it sets lakeFSAuthToken.  And every time after that !needsNewToken, so it does nothing.

Reply:
I can see very clearly see why it's confusing sorry for that. 

1. Regardless of implementation which we can discuss - the goal of this function is to create a new token (with new TTL) if possible. 
1. I'm trying to make a standard for calling refresh token, mimicking how [AWSCredentialProvider](https://github.com/aws/aws-sdk-java/blob/1.7.4/src/main/java/com/amazonaws/auth/AWSCredentialsProvider.java#L44) did that.
1. I'm open for discussion on removing it for the initial release - e.g no refresh will be delivered.

Reply:
So please document what it does: it clearly confuses at least one reader of the code.